File: appcore/src/main/java/mobi/cangol/mobile/utils/DataCleanManager.java
package mobi.cangol.mobile.utils;

import android.content.Context;
import android.os.Environment;

import java.io.File;

import mobi.cangol.mobile.logging.Log;

/**
 * 本应用数据清除管理器
 */
public class DataCleanManager {

    public static final String DATA_DATA = "/data/data/";
    public static final String SHARED_PREFS = "/shared_prefs";
    public static final String DATABASES = "/databases";

    private DataCleanManager() {
    }

    /**
     * 清除本应用内部缓存(/data/data/com.xxx.xxx/cache) * * @param context
     */
    public static void cleanInternalCache(Context context) {
        deleteFilesByDirectory(context.getCacheDir());
    }

    /**
     * 清除本应用所有数据库(/data/data/com.xxx.xxx/databases) * * @param context
     */
    public static void cleanDatabases(Context context) {
        deleteFilesByDirectory(new File(DATA_DATA + context.getPackageName() + DATABASES));
    }

    /**
     * * 清除本应用SharedPreference(/data/data/com.xxx.xxx/shared_prefs) * * @param
     * context
     */
    public static void cleanSharedPreference(Context context) {
        deleteFilesByDirectory(new File(DATA_DATA + context.getPackageName() + SHARED_PREFS));
    }

    /**
     * 按名字清除本应用数据库 * * @param context * @param dbName
     */
    public static void cleanDatabaseByName(Context context, String dbName) {
        context.deleteDatabase(dbName);
    }

    /**
     * 清除/data/data/com.xxx.xxx/files下的内容 * * @param context
     */
    public static void cleanFiles(Context context) {
        deleteFilesByDirectory(context.getFilesDir());
    }

    /**
     * * 清除外部cache下的内容(/mnt/sdcard/android/data/com.xxx.xxx/cache) * * @param
     * context
     */
    public static void cleanExternalCache(Context context) {
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            deleteFilesByDirectory(context.getExternalCacheDir());
        }
    }

    /**
     * 清除自定义路径下的文件，使用需小心，请不要误删。而且只支持目录下的文件删除 * * @param filePath
     */
    public static void cleanCustomCache(String filePath) {
        deleteFilesByDirectory(new File(filePath));
    }

    /**
     * 清除本应用所有的数据 * * @param context * @param filepath
     */
    public static void cleanApplicationData(Context context, String... filepath) {
        cleanInternalCache(context);
        cleanExternalCache(context);
        cleanDatabases(context);
        cleanSharedPreference(context);
        cleanFiles(context);
        for (final String filePath : filepath) {
            cleanCustomCache(filePath);
        }
    }

    /**
     * 删除方法 这里只会删除某个文件夹下的文件，如果传入的directory是个文件夹，将递归删除文件 * * @param directory
     */
    private static void deleteFilesByDirectory(File directory) {
        Log.d("deleteFilesByDirectory=" + directory);
        if (directory != null && directory.exists() && directory.isDirectory()) {
            for (final File item : directory.listFiles()) {
                if (!item.isDirectory()) {
                    final boolean result = item.delete();
                    Log.d("delete " + item.getAbsoluteFile() + ",result=" + result);
                } else {
                    deleteFilesByDirectory(item);
                }
            }
        }
    }

    public static long getFolderSize(File file) {
        long size = 0;
        for (final File item : file.listFiles()) {
            if (!item.isDirectory()) {
                size = size + getFolderSize(item);
            } else {
                size = size + item.length();
            }
        }
        return size;
    }

    public static long getAllCacheSize(Context context, String... filepath) {
        long size = 0;
        size = size + getFolderSize(context.getCacheDir());
        size = size + getFolderSize(context.getExternalCacheDir());
        size = size + getFolderSize(context.getFilesDir());
        size = size + getFolderSize(new File(DATA_DATA + context.getPackageName() + SHARED_PREFS));
        size = size + getFolderSize(new File(DATA_DATA + context.getPackageName() + DATABASES));
        for (final String filePath : filepath) {
            size = size + getFolderSize(new File(filePath));
        }
        return size;
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/utils/DataCleanManager.kt
package mobi.cangol.mobile.utils

import android.content.Context
import android.os.Environment
import mobi.cangol.mobile.logging.Log
import java.io.File

/**
 * 本应用数据清除管理器
 */
object DataCleanManager {

    private const val DATA_DATA = "/data/data/"
    private const val SHARED_PREFS = "/shared_prefs"
    private const val DATABASES = "/databases"

    /**
     * 清除本应用内部缓存(/data/data/com.xxx.xxx/cache) * * @param context
     */
    @JvmStatic
    fun cleanInternalCache(context: Context) {
        deleteFilesByDirectory(context.cacheDir)
    }

    /**
     * 清除本应用所有数据库(/data/data/com.xxx.xxx/databases) * * @param context
     */
    @JvmStatic
    fun cleanDatabases(context: Context) {
        deleteFilesByDirectory(File(DATA_DATA + context.packageName + DATABASES))
    }

    /**
     * * 清除本应用SharedPreference(/data/data/com.xxx.xxx/shared_prefs) * * @param
     * context
     */
    @JvmStatic
    fun cleanSharedPreference(context: Context) {
        deleteFilesByDirectory(File(DATA_DATA + context.packageName + SHARED_PREFS))
    }

    /**
     * 按名字清除本应用数据库 * * @param context * @param dbName
     */
    @JvmStatic
    fun cleanDatabaseByName(context: Context, dbName: String) {
        context.deleteDatabase(dbName)
    }

    /**
     * 清除/data/data/com.xxx.xxx/files下的内容 * * @param context
     */
    @JvmStatic
    fun cleanFiles(context: Context) {
        deleteFilesByDirectory(context.filesDir)
    }

    /**
     * * 清除外部cache下的内容(/mnt/sdcard/android/data/com.xxx.xxx/cache) * * @param
     * context
     */
    @JvmStatic
    fun cleanExternalCache(context: Context) {
        if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            deleteFilesByDirectory(context.externalCacheDir)
        }
    }

    /**
     * 清除自定义路径下的文件，使用需小心，请不要误删。而且只支持目录下的文件删除 * * @param filePath
     */
    @JvmStatic
    fun cleanCustomCache(filePath: String) {
        deleteFilesByDirectory(File(filePath))
    }

    /**
     * 清除本应用所有的数据 * * @param context * @param filepath
     */
    @JvmStatic
    fun cleanApplicationData(context: Context, vararg filepath: String) {
        cleanInternalCache(context)
        cleanExternalCache(context)
        cleanDatabases(context)
        cleanSharedPreference(context)
        cleanFiles(context)
        for (filePath in filepath) {
            cleanCustomCache(filePath)
        }
    }

    /**
     * 删除方法 这里只会删除某个文件夹下的文件，如果传入的directory是个文件夹，将递归删除文件 * * @param directory
     */
    private fun deleteFilesByDirectory(directory: File?) {
        if (directory != null && directory.exists() && directory.isDirectory) {
            for (item in directory.listFiles()) {
                if (!item.isDirectory) {
                    val result = item.delete()
                    Log.d("delete " + item.absoluteFile + ",result=" + result)
                } else {
                    deleteFilesByDirectory(item)
                }
            }
        }
    }

    @JvmStatic
    fun getFolderSize(file: File): Long {
        var size: Long = 0
        for (item in file.listFiles()) {
            size += if (!item.isDirectory) {
                getFolderSize(item)
            } else {
                item.length()
            }
        }
        return size
    }

    @JvmStatic
    fun getAllCacheSize(context: Context, vararg filepath: String): Long {
        var size: Long = 0
        size += getFolderSize(context.cacheDir)
        size += getFolderSize(context.externalCacheDir!!)
        size += getFolderSize(context.filesDir)
        size += getFolderSize(File(DATA_DATA + context.packageName + SHARED_PREFS))
        size += getFolderSize(File(DATA_DATA + context.packageName + DATABASES))
        for (filePath in filepath) {
            size += getFolderSize(File(filePath))
        }
        return size
    }
}